src/testrender/shading.cpp

```diff
@@ -433,24 +275,25 @@ register_closures(OSL::ShadingSystem* shadingsys)
     for (const BuiltinClosures& b : builtins)
         shadingsys->register_closure(b.name, b.id, b.params, nullptr, nullptr);
 }
+#endif // !defined(__CUDACC__)
 
 OSL_NAMESPACE_EXIT
 
 namespace {  // anonymous namespace
 
 template<int trans> struct Diffuse final : public BSDF, DiffuseParams {
-    Diffuse(const DiffuseParams& params) : BSDF(), DiffuseParams(params)
+    OSL_HOSTDEVICE Diffuse(const DiffuseParams& params) : BSDF(DIFFUSE_ID), DiffuseParams(params)
     {
         if (trans)
             N = -N;
     }
-    Sample eval(const Vec3& /*wo*/, const OSL::Vec3& wi) const override
+    OSL_HOSTDEVICE Sample eval(const Vec3& /*wo*/, const OSL::Vec3& wi) const OSL_HOSTDEVICE_OVERRIDE
     {
         const float pdf = std::max(N.dot(wi), 0.0f) * float(M_1_PI);
         return { wi, Color3(1.0f), pdf, 1.0f };
     }
-    Sample sample(const Vec3& /*wo*/, float rx, float ry,
-                  float /*rz*/) const override
+    OSL_HOSTDEVICE Sample sample(const Vec3& /*wo*/, float rx, float ry,
+                  float /*rz*/) const OSL_HOSTDEVICE_OVERRIDE
```

Comment:
Does this OSL_HOSTDEVICE_OVERRIDE mean you are keeping the virtual calls on CPU?

Reply:
Hmm, good catch. I'll get rid of the virtual calls.